File: src/Monitor.App/Services/TimerService.cs
using Monitor.App.Models;
using Monitor.App.Repositories;

namespace Monitor.App.Services;

public class TimerService
{
    private readonly IUsageRepository _usageRepo;
    private readonly IAuditRepository _auditRepo;

    private bool _running;
    private string _userName = "";
    private DateTime _startTimestamp;
    private long _startCounter; // Stopwatch.GetTimestamp()

    public bool IsRunning => _running;
    public string CurrentUserName => _running ? _userName : "";

    public TimerService(IUsageRepository usageRepo, IAuditRepository auditRepo)
    {
        _usageRepo = usageRepo;
        _auditRepo = auditRepo;
    }

    public bool StartTimer(string userName)
    {
        if (string.IsNullOrWhiteSpace(userName)) return false;
        _userName = userName.Trim();
        _startTimestamp = DateTime.Now;
        _startCounter = System.Diagnostics.Stopwatch.GetTimestamp();
        _running = true;
        _auditRepo.LogAction("timer_start", $"user={_userName}");
        return true;
    }

    public TimerStopResult? StopTimer()
    {
        if (!_running) return null;
        _running = false;

        var endTime = DateTime.Now;
        var elapsed = (long)((System.Diagnostics.Stopwatch.GetTimestamp() - _startCounter)
            / (double)System.Diagnostics.Stopwatch.Frequency);
        var durationSeconds = (int)elapsed;
        var userName = _userName;

        _usageRepo.AddUsageRecord(userName,
            _startTimestamp.ToString("yyyy-MM-dd HH:mm:ss"),
            endTime.ToString("yyyy-MM-dd HH:mm:ss"));

        _auditRepo.LogAction("timer_stop", $"user={userName},duration={durationSeconds}s");

        return new TimerStopResult
        {
            Success = true,
            UserName = userName,
            DurationSeconds = durationSeconds,
            DurationFormatted = FormatDuration(durationSeconds),
            ElapsedFormatted = FormatDuration(durationSeconds)
        };
    }

    public TimerState GetState()
    {
        if (!_running)
            return new TimerState();

        var elapsed = (int)((System.Diagnostics.Stopwatch.GetTimestamp() - _startCounter)
            / (double)System.Diagnostics.Stopwatch.Frequency);

        return new TimerState
        {
            IsRunning = true,
            CurrentUser = _userName,
            StartTime = _startTimestamp.ToString("yyyy-MM-dd HH:mm:ss"),
            ElapsedSeconds = elapsed,
            ElapsedFormatted = FormatDuration(elapsed)
        };
    }

    public List<UseHistoryItem> GetHistory(string filterMode = "all", string search = "")
    {
        var records = _usageRepo.GetUsageRecords(filterMode);
        if (!string.IsNullOrWhiteSpace(search))
            records = records.Where(r => r.UserName.Contains(search, StringComparison.OrdinalIgnoreCase)).ToList();
        return records;
    }

    public void DeleteRecord(int id)
    {
        _usageRepo.DeleteUsageRecord(id);
        _auditRepo.LogAction("record_deleted", $"id={id}");
    }

    public int DeleteRecords(List<int> ids)
    {
        _usageRepo.DeleteUsageRecords(ids);
        _auditRepo.LogAction("records_deleted", $"count={ids.Count}");
        return ids.Count;
    }

    public static string FormatDuration(int seconds)
    {
        var h = seconds / 3600;
        var m = (seconds % 3600) / 60;
        var s = seconds % 60;
        return $"{h:D2}:{m:D2}:{s:D2}";
    }
}

public class TimerStopResult
{
    public bool Success { get; set; }
    public string UserName { get; set; } = "";
    public int DurationSeconds { get; set; }
    public string DurationFormatted { get; set; } = "";
    public string ElapsedFormatted { get; set; } = "";
}

File: legacy/python-monitor/app/domain/services/timer_service.py
"""
Timer service — single source of truth for timer state and usage record persistence.
TimerFrame subscribes to this service for display; no dual state.
"""
import time
from datetime import datetime
from app.domain.repositories.usage_repository import UsageRepository
from app.domain.repositories.audit_repository import AuditRepository


class TimerService:
    """Manages timer lifecycle. The single authoritative source of timer state."""

    def __init__(self, usage_repo: UsageRepository, audit_repo: AuditRepository):
        self._usage_repo = usage_repo
        self._audit_repo = audit_repo
        self._reset_state()

    # ---- public state (read by TimerFrame and StatusService) ----
    @property
    def is_running(self) -> bool:
        return self._running

    @property
    def current_user_name(self) -> str:
        return self._user_name

    @property
    def start_timestamp(self):
        return self._start_timestamp

    @property
    def start_counter(self) -> float:
        return self._start_counter

    # ---- lifecycle ----

    def _reset_state(self):
        self._running = False
        self._user_name = ""
        self._start_timestamp = None
        self._start_counter = 0.0

    def start_timer(self, user_name: str) -> bool:
        """Start timing for a user. Returns False if name is empty.
        Caller (TimerFrame) should call this AFTER any entry animation completes,
        so the recorded start time matches the display start time.
        """
        name = user_name.strip() if user_name else ""
        if not name:
            return False
        self._user_name = name
        self._start_timestamp = datetime.now()
        self._start_counter = time.time()
        self._running = True
        self._audit_repo.log_action("timer_start", f"{name} 开始计时")
        return True

    def stop_timer(self) -> dict:
        """Stop timing, persist the usage record. Returns result dict or None.
        This is the ONLY path to persist a completed usage record.
        """
        if not self._running:
            return None
        end_dt = datetime.now()
        sec = int((end_dt - self._start_timestamp).total_seconds())
        self._usage_repo.add_usage_record(self._user_name, self._start_timestamp, end_dt)

        h, r = divmod(sec, 3600)
        m, s = divmod(r, 60)
        result = {
            "user_name": self._user_name,
            "duration_seconds": sec,
            "duration_formatted": f"{h:02}小时{m:02}分{s:02}秒",
            "elapsed_formatted": f"{h:02}:{m:02}:{s:02}",
        }
        self._audit_repo.log_action("timer_stop",
                                     f"{self._user_name} 停止计时, 时长{sec}秒")
        self._reset_state()
        return result

    def get_state(self) -> dict:
        """Get current timer state (for GUI display and web API)."""
        if not self._running:
            return {
                "is_running": False,
                "current_user": None,
                "start_time": None,
                "elapsed_seconds": 0,
                "elapsed_formatted": "00:00:00",
            }
        elapsed = int(time.time() - self._start_counter)
        h, r = divmod(elapsed, 3600)
        m, s = divmod(r, 60)
        return {
            "is_running": True,
            "current_user": self._user_name,
            "start_time": self._start_timestamp.strftime("%Y-%m-%d %H:%M:%S"),
            "elapsed_seconds": elapsed,
            "elapsed_formatted": f"{h:02}:{m:02}:{s:02}",
        }

    # ---- History / record management ----

    def get_history(self, filter_mode: str = "all", filter_text: str = "") -> list:
        """Get usage records, optionally filtered by name."""
        records = self._usage_repo.get_usage_records(filter_mode)
        if filter_text:
            records = [r for r in records if filter_text.lower() in r["user_name"].lower()]
        return records

    def delete_record(self, record_id: int):
        """Delete a single usage record."""
        self._usage_repo.delete_usage_record(record_id)
        self._audit_repo.log_action("delete_record", f"删除记录 ID: {record_id}")

    def delete_records(self, record_ids: list):
        """Delete multiple usage records."""
        if not record_ids:
            return
        self._usage_repo.delete_usage_records(record_ids)
        self._audit_repo.log_action("delete_records", f"批量删除 {len(record_ids)} 条记录")
